server/api/utils/clients/async_nuclio.py

```diff
@@ -0,0 +1,105 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import copy
+import urllib.parse
+
+import aiohttp
+
+import mlrun.common.schemas
+import mlrun.utils
+from mlrun.utils import logger
+
+NUCLIO_API_SESSIONS_ENDPOINT = "/api/sessions/"
+NUCLIO_API_GATEWAYS_ENDPOINT = "/api/api_gateways/"
+API_GATEWAY_NAMESPACE_HEADER = "X-Nuclio-Api-Gateway-Namespace"
+NUCLIO_PROJECT_NAME_HEADER = "X-Nuclio-Project-Name"
+
+
+class Client:
+    def __init__(self, auth_info: mlrun.common.schemas.AuthInfo):
+        self._async_session = None
+        self._auth = aiohttp.BasicAuth(auth_info.username, auth_info.session)
+        self._logger = logger.get_child("nuclio-client")
+        self.nuclio_dashboard_url = mlrun.mlconf.nuclio_dashboard_url
+
+    async def _ensure_async_session(self):
+        if not self._async_session:
+            self._async_session = mlrun.utils.AsyncClientWithRetry(
+                retry_on_exception=mlrun.mlconf.httpdb.projects.retry_leader_request_on_exception
+                == mlrun.common.schemas.HTTPSessionRetryMode.enabled.value,
+                logger=logger,
+            )
+
+    async def _send_request_to_api(
+        self, method, url, path="/", error_message: str = "", **kwargs
+    ):
+        await self._ensure_async_session()
+        response = await self._async_session.request(
+            method=method,
+            url=urllib.parse.urljoin(url, path),
+            verify_ssl=False,
+            auth=self._auth,
+            **kwargs,
+        )
+        if not response:
+            return
+        if not response.ok:
+            try:
+                response_body = await response.json()
+            except Exception:
+                response_body = {}
+            self._handle_error_response(
+                method, path, response, response_body, error_message, kwargs
+            )
+        else:
+            return await response.json()
+
+    def _handle_error_response(
+        self, method, path, response, response_body, error_message, kwargs
+    ):
+        log_kwargs = copy.deepcopy(kwargs)
+
+        # this can be big and spammy
+        log_kwargs.pop("json", None)
+        log_kwargs.update({"method": method, "path": path})
+        try:
+            ctx = response_body.get("meta", {}).get("ctx")
+            errors = response_body.get("errors", [])
+        except Exception:
+            pass
+        else:
+            if errors:
+                error_message = f"{error_message}: {str(errors)}"
+            if errors or ctx:
+                log_kwargs.update({"ctx": ctx, "errors": errors})
```

Comment:
I am not sure this handles nuclio error response structure correctly as nuclio does not return meta/ctx and co.

Reply:
@liranbg fixed to handling more nuclio specific. Also, in case of any not nuclio-related issue middleware will raise exceptions before.